Add time-variable fares examples

Sergiodero opened the pull request:
[A vote to add time-variable fares has passed](https://github.com/google/transit/pull/357). These changes add two different use cases on the Fares v2 example page to illustrate the use of time-variable fares on its own (Washington DC example), and in combination with zone-based fares (MTA Metro-North example).

Sergiodero committed:
Add time-variable fares example

Added Washington DC Metrorail example

Sergiodero committed:
Add time-variable + zone based fares

Added MTA Metro-North example illustrating the combined use of time variable fares and zone based fares.

emmambd commented on docs/schedule/examples/fares-v2.md:
```diff
@@ -216,7 +216,167 @@ The single ride fare product shown below has both `cash` and `tap-to-ride` fare
 
+Some transit agencies differentiate fares depending on the time and/or day of the week, where fares can be associated with specific time periods such as a peak, off-peak and/or weekends. 
 
+Washington DC’s <a href="https://www.wmata.com/fares/basic.cfm" target="_blank">Metrorail fares</a> vary based on multiple factors, including day and time of trip. Variable time fares in GTFS can be defined using `timeframes.txt`, in which it is possible to designate specific time periods that then can be associated in `fare_leg_rules.txt` to assign the applicable fare product that corresponds to ...
```
it could be slightly clearer that this is hypothetical (just so people don't mistakenly assume WMATA is in fact using time variable fares right now in their feed).

isabelle-dr replied:
day and time of *the* trip?

To @emmambd's point, we could add a sentence saying" The following example is fictional, based on WMATA's fares."

Sergiodero replied:
The text has been updated to include both the "day and time of the trip" correction and the clarification that this is a fictional example

isabelle-dr commented on docs/schedule/examples/fares-v2.md:
```diff
@@ -216,7 +216,167 @@ The single ride fare product shown below has both `cash` and `tap-to-ride` fare
 <sup><a href="https://gtfs.calitp.org/production/CleanAirExpressFaresv2.zip" target="_blank">Download the Clean Air Express feed</a></sup>
 
 
+## Define price differences based on time of trip
```
I would have "time and day" because days always come into play (where time sometimes doesn't)

isabelle-dr commented on docs/schedule/examples/fares-v2.md:
```diff
@@ -216,7 +216,167 @@ The single ride fare product shown below has both `cash` and `tap-to-ride` fare
 
+## Define price differences based on time of trip
 
+Some transit agencies differentiate fares depending on the time and/or day of the week, where fares can be associated with specific time periods such as a peak, off-peak and/or weekends. 
```
Suggestion for clarity:

Certain transit agencies vary their fares based on the time and/or day of the week. This means that fares are associated with a time period where the trip is made, such as peak, off-peak hours, or weekends.

isabelle-dr commented on docs/schedule/examples/fares-v2.md:
```diff
@@ -216,7 +216,167 @@ The single ride fare product shown below has both `cash` and `tap-to-ride` fare
 
+Washington DC’s <a href="https://www.wmata.com/fares/basic.cfm" target="_blank">Metrorail fares</a> vary based on multiple factors, including day and time of trip. Variable time fares in GTFS can be defined using `timeframes.txt`, in which it is possible to designate specific time periods that then can be associated in `fare_leg_rules.txt` to assign the applicable fare product that corresponds to ...
+
+First, service days are defined using `calendar.txt`.
```
It would be useful to have somewhere (probably better written):

It is possible to create new records with `service_id` values that are not assigned to any trip and are only used for fares.

isabelle-dr replied:
I saw that you've added it below in the second example :) 
It might be overkill to mention it here as well, I'll leave it up to you

Sergiodero replied:
To keep thing simple in this initial example I'll leave it as is :)

isabelle-dr commented on docs/schedule/examples/fares-v2.md:
```diff
@@ -216,7 +216,167 @@ The single ride fare product shown below has both `cash` and `tap-to-ride` fare
+| sunday_service   | 0      | 0       | 0         | 0        | 0      | 0        | 1      | 20220708   | 20221231 |
+
+
+Afterwards, the desired timeframes are defined, providing an id, start time and end time for each time period, and associating each one of them with an entry from the field `calendar.service_id`.
```
Suggestion, because time periods are not always there but days always are:

Afterwards, the desired timeframes are defined in `timeframes.txt`, providing an id, the applicable days via a reference to `calendar.service_id`, and if applicable, the start time and end time for each time period.

isabelle-dr commented on docs/schedule/examples/fares-v2.md:
```diff
@@ -216,7 +216,167 @@ The single ride fare product shown below has both `cash` and `tap-to-ride` fare
+| weekend            |            |          | saturday_service |
+| weekend            |            |          | sunday_service   |
+
+Next, the corresponding time specific fares in fare_products.txt are created (e.g. Peak fare)
```
in `fare_products.txt`

isabelle-dr commented on docs/schedule/examples/fares-v2.md:
```diff
@@ -216,7 +216,167 @@ The single ride fare product shown below has both `cash` and `tap-to-ride` fare
+| weekend_fare    | Weekend Metrorail one-way fare                | 2      | USD      |
+| late_night_fare | Late Night flat fare (Mon - Fri after 9:30pm) | 2      | USD      |
+
+Finally both of these datasets are associated in `fare_leg_rules.txt` using the fields `from_timeframe_group_id` and `to_timeframe_group_id` to assign a `fare_product_id` to each specific combination of timeframes, or in this case, only using `from_timeframe_group_id` as the fare product depends only on the departure timeframe, leaving `to_timeframe_group_id` blank. 
```
Suggestion to add a bit more clarify the intended use of the "start" and "end" fields. 

Lastly, timeframes are associated with fare products in `fare_leg_rules.txt` using the fields `from_timeframe_group_id` and `to_timeframe_group_id`. These fields determine whether a fare applies solely to the start of the leg or both the start and end of the leg.
For WMATA, the fare depends only on the leg's departure timeframe, so  `to_timeframe_group_id` is left blank.

isabelle-dr commented on docs/schedule/examples/fares-v2.md:
```diff
@@ -216,7 +216,167 @@ The single ride fare product shown below has both `cash` and `tap-to-ride` fare
+| 1          | peak_fare       | weekday_peak            |                       | (fare gate)    |
+| 1          | regular_fare    | weekday_offpeak         |                       | (fare gate)    |
+
+Note that `network_id references` the foreign ID: `routes.network_id`, and that arrival and departure information on `stop_times.txt` along with `timeframes.txt` will inform the selection of the correct fare_product for each specific trip. 
```
Suggestion: small formatting change reformulating second portion

Note that `network_id` references the foreign ID `routes.network_id`, and that the selection of the correct fare product for each trip will be a combination of arrival and departure times from `stop_times.txt` along with times defined in `timeframes.txt.

isabelle-dr commented on docs/schedule/examples/fares-v2.md:
```diff
@@ -216,7 +216,167 @@ The single ride fare product shown below has both `cash` and `tap-to-ride` fare
+
+## Define time-variable fares along with zone based fares
+
+In New York's MTA Metro-North railroad network, fares vary based on both the time of the day and the areas of the trips’s origin and destination. The following example illustrates the fare rules applicable to a trip from Grand Central Station to Poughkeepsie (NY, USA).
```
Suggestion:

In New York's MTA Metro-North railroad network, fares vary based on both the time of the day and the trips’s origin and destination areas.

isabelle-dr commented on docs/schedule/examples/fares-v2.md:
```diff
@@ -216,7 +216,167 @@ The single ride fare product shown below has both `cash` and `tap-to-ride` fare
+
+In New York's MTA Metro-North railroad network, fares vary based on both the time of the day and the areas of the trips’s origin and destination. The following example illustrates the fare rules applicable to a trip from Grand Central Station to Poughkeepsie (NY, USA).
+
+This example based on a <a href="https://docs.google.com/spreadsheets/d/1-cD-R2OH5xAQAbNWNlrXD7WOw594lVdW-bomuLo6bI8/edit?usp=sharing" target="_blank">dataset</a> produced by <a href="https://www.itoworld.com/" target="_blank">ITO World</a>, features a trip that uses ten stops distributed in six different areas.
```
features -> featuring

isabelle-dr commented on docs/schedule/examples/fares-v2.md:
```diff
@@ -216,7 +216,167 @@ The single ride fare product shown below has both `cash` and `tap-to-ride` fare
+| mnr_HUD-9 | ITO2096 |
+
+
+Service days for train services 3 and 13 are defined using calendar.txt. Notably, general service days (i.e. weekdays, weekends and anyday) are also defined, as these will be associated with specific timeframes in order to model `time-variable fares`.
```
Suggestion, I don't think we should use "service days" for describing days that aren't associated with a trip. 

Service days for train services 3 and 13 are defined using `calendar.txt`. Notably, other records with generic days (i.e. weekdays, weekends, and anyday) that aren't associated with any trips are defined, and these will be associated with timeframes in order to model time-variable fares.

isabelle-dr commented on docs/schedule/examples/fares-v2.md:
```diff
@@ -216,7 +216,167 @@ The single ride fare product shown below has both `cash` and `tap-to-ride` fare
+| anyday     | 1      | 1       | 1         | 1        | 1      | 1        | 1      | 20220101   | 20240101 |
+
+
+Specific time frames are created in `timeframes.txt`, including general service days (anytime, weekdays and weekends), and peak and off peak periods:
```
Suggestion to avoid using "general service days" which could be confusing:

Records are created in `timeframes.txt`, including cases where the time covers the 24-hour range period (`anytime`, `weekdays` and `weekends`), and peak and off-peak periods:

isabelle-dr commented on docs/schedule/examples/fares-v2.md:
```diff
@@ -216,7 +216,167 @@ The single ride fare product shown below has both `cash` and `tap-to-ride` fare
+| mnr_notam2pmpeak   | 20:00:00   | 24:00:00 | weekdays   |
+
+
+Each individual fare product is defined in `fare_products.txt`, in this example, only trips between zone 1 and 9 are listed. The full dataset would include a record for each  price defined by a time and zone combination. Additionally, the example only uses one fare media (i.e. paper ticket) but additional combinations could be created if prices would also vary depending on fare media.
```
Suggestion

Each individual fare product is defined in `fare_products.txt`. In this example, only trips between zone 1 and 9 are listed. The full dataset would include a record for each price defined by a time and zone combination. Additionally, the example only displays one fare media (`paper`) but additional combinations could be created if prices would also vary based on the fare media.

isabelle-dr commented on docs/schedule/examples/fares-v2.md:
```diff
@@ -216,7 +216,167 @@ The single ride fare product shown below has both `cash` and `tap-to-ride` fare
+| mnr_HUD-9:1_adult      | Inbound Adult Off Peak Zonal Fare  | paper         | 19.25  | USD      |
+
+
+Finally, origin and destination areas and timeframes combinations are associated with the corresponding fare product in `fare_leg_rules.txt`. Here, trips starting or arriving in Zone 1 (i.e. `area_id=mnr_1`) during peak times are subject to a specific peak fare corresponding to the arrival and departure zones of the trip (i.e. `fare_product_id=mnr_1:HUD-9_adult_peak`).
```
Recommendation (I think in English, "finally" might be more "I finally got to experience the REM" type of sentences, but I might be wrong as I am not a native English speaker!)

Replacing:

> Finally, origin and destination areas and timeframes combinations are associated with the corresponding fare product in `fare_leg_rules.txt`.


With:

> Lastly, the combinations of origin and destination areas, along with their respective timeframes are associated with ...

Sergiodero replied:
Thanks for this! I really appreciate those checks as I'm also not a native English speaker :)

isabelle-dr reviewed (commented):
Thank you for working on this @Sergiodero!

This is a great addition to the examples page and it makes it very clear how to use timeframes.
I've added a few suggestions intended to slightly increase the clarity. Feel free to only take what you think is an improvement. 

Another thing I wonder if we should make an explicit mention is that timeframes are using local time semantics, with a link to [this section of the spec](https://gtfs.org/schedule/reference/#timeframe-local-time-semantics), because this is a new concept introduced.

Sergiodero committed:
Update to fares-v2 examples to address feedback from review

Resolves comments from first round of reviews on time-variable fares examples.

Sergiodero commented:
Thanks for the review @isabelle-dr! All comments have been addressed in the latest commit. 

I'll work on a new commit now to update the NYC example so that it matches the new screenshots to illustrate the example.

Sergiodero committed:
Update time-variable + zones fares example

Replaced time-variable fares + zone based fares example with trip to Cold Spring, NY. Added Apple Maps screenshots.

Sergiodero committed:
Added time-variable fares screen captures

Sergiodero committed:
Updated links to time-variable fare images

Sergiodero commented:
Last updates now reflect changes made to the time-variable + zonal fares example and include screen captures from Apple Maps.

isabelle-dr reviewed (approved):
Thank you for updating this, LGTM!
